src/main/kotlin/com/terraformation/backend/nursery/db/BatchImporter.kt

```diff
@@ -0,0 +1,126 @@
+import com.terraformation.backend.db.default_schema.UploadId
+import com.terraformation.backend.db.default_schema.UploadType
+import com.terraformation.backend.db.default_schema.tables.daos.UploadProblemsDao
+import com.terraformation.backend.db.default_schema.tables.daos.UploadsDao
+import com.terraformation.backend.db.default_schema.tables.pojos.SpeciesRow
+import com.terraformation.backend.db.default_schema.tables.pojos.UploadsRow
+import com.terraformation.backend.db.nursery.tables.pojos.BatchesRow
+import com.terraformation.backend.file.FileStore
+import com.terraformation.backend.file.UploadService
+import com.terraformation.backend.file.UploadStore
+import com.terraformation.backend.i18n.Messages
+import com.terraformation.backend.importer.CsvImporter
+import com.terraformation.backend.importer.CsvValidator
+import com.terraformation.backend.species.db.SpeciesStore
+import java.io.InputStream
+import java.time.LocalDate
+import javax.annotation.ManagedBean
+import org.jobrunr.scheduling.JobScheduler
+import org.jooq.DSLContext
+import org.springframework.context.annotation.Lazy
+
+@ManagedBean
+class BatchImporter(
+    private val batchStore: BatchStore,
+    dslContext: DSLContext,
+    fileStore: FileStore,
+    private val messages: Messages,
+    private val parentStore: ParentStore,
+    @Lazy scheduler: JobScheduler,
+    private val speciesStore: SpeciesStore,
+    uploadProblemsDao: UploadProblemsDao,
+    uploadsDao: UploadsDao,
+    uploadService: UploadService,
+    uploadStore: UploadStore,
+    userStore: UserStore,
+) :
+    CsvImporter(
+        dslContext,
+        fileStore,
+        scheduler,
+        uploadProblemsDao,
+        uploadsDao,
+        uploadService,
+        uploadStore,
+        userStore) {
+  override val templatePath: String
+    get() = "/csv//batches-template.csv"
+
+  fun receiveCsv(inputStream: InputStream, fileName: String, facilityId: FacilityId): UploadId {
+    requirePermissions { createBatch(facilityId) }
+
+    if (parentStore.getFacilityType(facilityId) != FacilityType.Nursery) {
+      throw FacilityTypeMismatchException(facilityId, FacilityType.Nursery)
+    }
+
+    val organizationId = parentStore.getOrganizationId(facilityId)
+    return doReceiveCsv(
+        inputStream, fileName, UploadType.SeedlingBatchCSV, organizationId, facilityId)
+  }
+
+  override fun getValidator(uploadsRow: UploadsRow): CsvValidator {
+    return BatchCsvValidator(uploadsRow.id!!, messages)
+  }
+
+  override fun doImportCsv(
+      uploadsRow: UploadsRow,
+      csvReader: CSVReader,
+      overwriteExisting: Boolean
+  ) {
+    val organizationId = uploadsRow.organizationId!!
+    val facilityId = uploadsRow.facilityId!!
+
+    requirePermissions {
+      createBatch(facilityId)
+      createSpecies(organizationId)
+    }
+
+    // Consume header row
+    csvReader.readNext()
+
+    dslContext.transaction { _ -> csvReader.forEach { importRow(it, organizationId, facilityId) } }
```

Comment:
if any single row fails, it will fail the whole transaction, correct? That is, it won't skip invalid rows and continue?

Reply:
> if any single row fails, it will fail the whole transaction, correct? That is, it won't skip invalid rows and continue?

Correct. We don't want to end up with a situation where the file is half-imported and the user has to go edit their CSV file to remove the successfully-imported lines before they try again.

Reply:
that makes sense!